presto-main/src/main/java/io/prestosql/sql/planner/DomainTranslator.java

```diff
@@ -857,6 +867,88 @@ protected ExtractionResult visitBetweenPredicate(BetweenPredicate node, Boolean
+        {
+            Optional<ExtractionResult> result = tryVisitLikePredicate(node, complement);
+            if (result.isPresent()) {
+                return result.get();
+            }
+            return super.visitLikePredicate(node, complement);
+        }
+
+        private Optional<ExtractionResult> tryVisitLikePredicate(LikePredicate node, Boolean complement)
+        {
+            if (!(node.getValue() instanceof SymbolReference)) {
+                // LIKE not on a symbol
+                return Optional.empty();
+            }
+
+            if (!(node.getPattern() instanceof StringLiteral)) {
+                // dyamic pattern
+                return Optional.empty();
+            }
+
+            if (node.getEscape().isPresent() && !(node.getEscape().get() instanceof StringLiteral)) {
+                // dynamic escape
+                return Optional.empty();
+            }
+
+            Type type = typeAnalyzer.getType(session, types, node.getValue());
+            if (!(type instanceof VarcharType)) {
+                // not varchar
+                return Optional.empty();
+            }
+            VarcharType varcharType = (VarcharType) type;
+
+            Symbol symbol = Symbol.from(node.getValue());
+            Slice pattern = ((StringLiteral) node.getPattern()).getSlice();
+            Optional<Slice> escape = node.getEscape()
+                    .map(StringLiteral.class::cast)
+                    .map(StringLiteral::getSlice);
+
+            int patternConstantPrefixBytes = LikeFunctions.patternConstantPrefixBytes(pattern, escape);
+            if (patternConstantPrefixBytes == pattern.length()) {
+                Slice literal = LikeFunctions.unescapeLiteralLikePattern(pattern, escape);
+                ValueSet valueSet;
+                if (varcharType.isUnbounded() || countCodePoints(literal) <= varcharType.getBoundedLength()) {
+                    valueSet = ValueSet.of(type, literal);
+                }
+                else {
+                    // impossible to satisfy
+                    valueSet = ValueSet.none(type);
+                }
+                Domain domain = Domain.create(complementIfNecessary(valueSet, complement), false);
+                return Optional.of(new ExtractionResult(TupleDomain.withColumnDomains(ImmutableMap.of(symbol, domain)), TRUE_LITERAL));
+            }
+
+            if (complement || patternConstantPrefixBytes == 0) {
+                // TODO
+                return Optional.empty();
+            }
+
+            Slice constantPrefix = LikeFunctions.unescapeLiteralLikePattern(pattern.slice(0, patternConstantPrefixBytes), escape);
+
+            int lastIncrementable = -1;
+            for (int position = 0; position < constantPrefix.length(); position += lengthOfCodePoint(constantPrefix, position)) {
+                // stay within ASCII to be on the safe side with connectors TODO remove this limitation
```

Comment:
This comment confused me at first reading. I thought it meant that the prefix had to be all ASCII, but the code below says otherwise (i.e., find the last ASCII character).

Reply:
reworded